### PR TITLE
CI pipelinine to autobuild for various OS & Archs

### PR DESCRIPTION
**Pull request by OemerYilmazConsiderIT**
linux-x86_x64, linux-arm64, linux-arm32, windows-amd64, windows-amd32, macos-intel64, macos-AppleSilicon

### CODE REVIEW

**Comment by OemerYilmazConsiderIT** [comment]
@jbeyerstedt can you double check, if the workflow file is correct ?

**Comment by OemerYilmazConsiderIT** [comment]
> @jbeyerstedt can you double check, if the workflow file is correct ?

I did try it out with a PR, "vx.x.x" Tag and push on main branch. Seems to work fine. Im not sure if the filenaming is "correct"

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
+name: Go
+
+on:
+  push:
+    branches: [ "v2" ]
```
Comment (jbeyerstedt) [inline .github/workflows/go.yml]: We should only build on tags on the main branch (which is `v2` in this case) and pull requests against it. So this line is not needed.

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            arch: arm64
+          - name: linux-arm32
+            os: linux
+            arch: arm
+          - name: windows-amd64
+            os: windows
+            arch: amd64
+          - name: windows-amd32
+            os: windows
+            arch: 386
```
Comment (jbeyerstedt) [inline .github/workflows/go.yml:40]: I don't think that we need Windows 32 bit.

File: .github/workflows/go.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+          - name: macos-intel64
+            os: darwin
+            arch: amd64
+          - name: macos-AppleSilicon
```
Comment (jbeyerstedt) [inline .github/workflows/go.yml:44]: Why not `macos-amd64` and `macos-arm64`?

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+          name: mqtt-record-replay-${{ matrix.name }}-${{ env.app_version }}
+          path: ${{ matrix.name }}-${{ env.app_version }}.zip
+
+      - uses: svenstaro/upload-release-action@v2
+        name: Upload binaries to release
+        if: ${{ startsWith(github.ref, 'refs/tags/') }}
+        with:
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          file: ${{ matrix.name }}-${{ env.app_version }}.zip
+          file_glob: true
```
Comment (jbeyerstedt) [inline .github/workflows/go.yml:86]: You don't need file glob

File: .github/workflows/go.yml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        with:
+          name: mqtt-record-replay-${{ matrix.name }}-${{ env.app_version }}
+          path: ${{ matrix.name }}-${{ env.app_version }}.zip
+
+      - uses: svenstaro/upload-release-action@v2
+        name: Upload binaries to release
+        if: ${{ startsWith(github.ref, 'refs/tags/') }}
+        with:
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          file: ${{ matrix.name }}-${{ env.app_version }}.zip
```
Comment (jbeyerstedt) [inline .github/workflows/go.yml:85]: I would like to have the application name in the release zip. Either use `asset_name` or already have the right name in the zipping step.

**Review by jbeyerstedt — CHANGES_REQUESTED** [review]

File: .github/workflows/go.yml (hunk 2 above)
Comment (OemerYilmazConsiderIT) [reply .github/workflows/go.yml:40]: ah okay, i agree. Windows 64Bit should be enough

File: .github/workflows/go.yml (hunk 3 above)
Comment (OemerYilmazConsiderIT) [reply .github/workflows/go.yml:44]: okay, ill change it

File: .github/workflows/go.yml (hunk 4 above)
Comment (OemerYilmazConsiderIT) [reply .github/workflows/go.yml:86]: ok

File: .github/workflows/go.yml (hunk 5 above)
Comment (OemerYilmazConsiderIT) [reply .github/workflows/go.yml:85]: @jbeyerstedt What do you mean with "application name" ? can you name an example ?

File: .github/workflows/go.yml (hunk 1 above)
Comment (OemerYilmazConsiderIT) [reply .github/workflows/go.yml]: ok

File: .github/workflows/go.yml (hunk 5 above)
Comment (jbeyerstedt) [reply .github/workflows/go.yml:85]: mqtt-record-replay-linux-arm64-$version.zip instead of just linux-arm64-$version.zip

**Comment by OemerYilmazConsiderIT** [comment]
@jbeyerstedt dont merge yet. There are sone issues it seems like

**Comment by OemerYilmazConsiderIT** [comment]
@jbeyerstedt Okay now its ready to be merged

File: .github/workflows/go.yml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      - name: Make build results executable
+        run: chmod +x mqtt-re*
+
+      - name: Zipping
+        run: zip -r mqtt-record-replay-${{ matrix.name }}-${{ env.app_version }}.zip ${{ matrix.name }}
+
+      - uses: actions/upload-artifact@v2
+        with:
+          name: mqtt-record-replay-${{ matrix.name }}-${{ env.app_version }}
+          path: mqtt-record-replay-${{ matrix.name }}-${{ env.app_version }}.zip
```
Comment (jbeyerstedt) [inline .github/workflows/go.yml:74]: upload-artifact already does a zip file, so please just specify the two file paths here.

**Review by jbeyerstedt — CHANGES_REQUESTED** [review]

File: .github/workflows/go.yml (hunk 6 above)
Comment (OemerYilmazConsiderIT) [reply .github/workflows/go.yml:74]: Okay. Good to know. I thought it would just upload files

Comment (jbeyerstedt) [reply .github/workflows/go.yml]: It's a bit stupid though as it does a double-zip now.

Comment (OemerYilmazConsiderIT) [reply .github/workflows/go.yml]: yes. im changing it

**Comment by jbeyerstedt** [comment]
Borked windows build in between because it's the only one with `.exe` file names.

**Review by jbeyerstedt — APPROVED** [review]
let's hope that it works with a release/ tag, too.

**Pull request merged**